### PR TITLE
Optional parameter changes in 8.1, and description improvements

### PR DESCRIPTION
**Pull request by IMSoP**
* Document change in behaviour of misplaced optional parameters in 8.1
* Expand and update documentation of default/optional parameters

### CODE REVIEW

**Comment by cronlabspl** [comment]
Lgtm

File: appendices/migration81/incompatible.xml [hunk 1]
```diff
@@ -55,6 +55,57 @@ var_dump(B::counter()); // int(4), previously int(2)
...
+catch (Error $e)
+{
+    echo get_class($e), ' - ', $e->getMessage(), "\n";
+}
+?>
+]]>
+     </programlisting>
+     &example.outputs.80;
+     <screen>
+      <![CDATA[
```
Comment (saundefined) [inline appendices/migration81/incompatible.xml:90]:
```suggestion
<![CDATA[
```

nit: indentation

Comment (cronlabspl) [reply appendices/migration81/incompatible.xml:90]: > nit: indentation

why line `90` shouldnt be indented?

Comment (cmb69) [reply appendices/migration81/incompatible.xml:90]: That's just an (undocumented) convention for the PHP manual sources.

File: appendices/migration81/incompatible.xml [hunk 2]
```diff
@@ -55,6 +55,57 @@ var_dump(B::counter()); // int(4), previously int(2)
...
+  <sect3 xml:id="migration81.incompatible.core.optional-before-required">
+   <title>Optional parameters specified before required parameters</title>
+
+   <para>
+    An <link linkend="functions.arguments.default">optional parameter</link>
+    specified before required parameters is now always treated as required,
+    even when called using
+    <link linkend="functions.named-arguments">named arguments</link>.
+    In PHP 8.0, the below emits a deprecation notice on the definition,
+    but runs successfully when called. As of PHP 8.1, an error is thrown.
```
Comment (cmb69) [inline appendices/migration81/incompatible.xml:67]: The convention is to always specify the full version (major, minor, revision), so:
```suggestion
    Prior to PHP 8.1.0, the below emits a deprecation notice on the definition,
    but runs successfully when called. As of PHP 8.1.0, an error is thrown.
```

File: language/functions.xml [hunk 3]
```diff
@@ -384,23 +410,62 @@ Making a bowl of raspberry .
...
+    <note>
+     <simpara>
+      As of PHP 7.1, omitting a parameter which does not specify a default
+      throws an <classname>ArgumentCountError</classname>; in previous versions
+      it raised a Warning.
+     </simpara>
+     <simpara>
+      As of PHP 8.0, <emphasis>defining</emphasis> a function with optional
+      parameters after required parameters (as in the first "yogurt"
+      example above) is DEPRECATED.
```
Comment (cmb69) [inline language/functions.xml:466]: Maybe better (I don't like screaming):
```suggestion
      example above) is <emphasis>deprecated</emphasis>.
```

**Review by cmb69 — COMMENTED** [review]
Thank you for the PR!

Maybe it's worthwhile to document that a `null` default still signals that the type is nullable (although the parameter would not be optional).

File: appendices/migration81/incompatible.xml (hunk 2 above)
Comment (IMSoP) [reply appendices/migration81/incompatible.xml:67]: The oddity here is that the code *only* runs successfully on 8.0.x, not any earlier versions, so "prior to" feels a bit misleading. I'll try to come up with an alternative wording.

Comment (cmb69) [reply appendices/migration81/incompatible.xml:67]: Oh, right. Maybe the clumsy "As of PHP 8.0.0, but prior to PHP 8.1.0" will do?

Comment (IMSoP) [reply appendices/migration81/incompatible.xml:67]: Yeah, I can't think of anything better at the moment.

File: language/functions.xml (hunk 3 above)
Comment (IMSoP) [reply language/functions.xml:466]: It turns out, this was already mentioned in an oddly placed paragraph right at the top of the page, so I've moved it down into this section.

File: appendices/migration81/incompatible.xml (hunk 1 above)
Comment (IMSoP) [reply appendices/migration81/incompatible.xml:90]: I had noticed that convention, but PhpStorm hadn't, so had "helpfully" indented things when I edited them. Should be fixed now.

File: language/functions.xml (hunk 3 above)
Comment (cronlabspl) [reply language/functions.xml]: > Maybe better (I don't like screaming):
> 
> ```suggestion
> 
>       example above) is <emphasis>deprecated</emphasis>.
> 
> ```

Now that's better😀

**Comment by cmb69** [comment]
It might be good to wait on the outcome of https://externals.io/message/116597#116643 before documenting the current behavior.

**Comment by cronlabspl** [comment]
@IMSoP Im perfectly OK with it, but dont have `write-access` here.........

**Comment by cmb69** [comment]
Thank you!

**Pull request merged**